kroxylicious-filters/kroxylicious-authorization/src/main/java/io/kroxylicious/filter/authorization/AuthorizationFilter.java

```diff
@@ -0,0 +1,278 @@
+        return cClass.cast(removed);
+    }
+
+    <R> R popAndApplyInflightState(ResponseHeaderData header, R response) {
+        var completer = popInflightState(header, InflightState.class);
+        if (completer != null) {
+            return (R) completer.merge(response);
+        }
+        else {
+            return response;
+        }
+    }
+
+    @Nullable
+    private CompletionStage<RequestFilterResult> checkRequestApiAndVersions(ApiKeys apiKey,
+                                                                            RequestHeaderData header,
+                                                                            ApiMessage request,
+                                                                            FilterContext context) {
+        if (!isApiVersionSupported(apiKey, header.requestApiVersion())) {
+            if (isApiSupported(apiKey)) {
+                LOG.warn("Filter of type {} does not support {} API version {} used in request."
+                        + " It supports version {} to {} (inclusive) of this API."
+                        + " This error is due to a misconfigured, buggy, or possibly malicious client.",
+                        getClass().getName(),
+                        apiKey,
+                        header.requestApiVersion(),
+                        minSupportedApiVersion(apiKey),
+                        maxSupportedApiVersion(apiKey));
+            }
+            else {
+                LOG.warn("Filter of type {} does not support {} API version {} used in request."
+                        + " It does not support version this API at all."
+                        + " This error is due to a misconfigured, buggy, or possibly malicious client.",
+                        getClass().getName(),
+                        apiKey,
+                        header.requestApiVersion());
+            }
+            return context.requestFilterResultBuilder()
+                    .errorResponse(header, request, Errors.UNSUPPORTED_VERSION.exception())
+                    .completed();
+        }
+        return null;
+    }
+
+    @Override
+    public CompletionStage<RequestFilterResult> onRequest(ApiKeys apiKey,
+                                                          RequestHeaderData header,
+                                                          ApiMessage request,
+                                                          FilterContext context) {
+
+        var usesUnsupportedApi = checkRequestApiAndVersions(apiKey,
+                header,
+                request,
+                context);
+        if (usesUnsupportedApi != null) {
+            return usesUnsupportedApi;
+        }
+
+        var enforcement = apiEnforcement.get(apiKey);
+        if (enforcement != null) {
+            return enforceRequest(header, request, context, enforcement);
+        }
+        else {
+            return context.forwardRequest(header, request);
+        }
```

Comment:
This impl feels a bit odd to read for a filter that fails closed. Yes I understand that to get to line 201 it has to be a supported API but I think that is obscuring a central component of the filter. 

 ```suggestion
        boolean supportedCombination = isApiVersionSupported(apiKey, header.requestApiVersion());
        if (supportedCombination) {
            var enforcement = apiEnforcement.get(apiKey);
            if (enforcement != null) {
                return enforceRequest(header, request, context, enforcement);
            }
            else {
                return context.forwardRequest(header, request);
            }
        }
        else {
            logUnsupportedVersion(apiKey, header);
            return context.requestFilterResultBuilder()
                    .errorResponse(header, request, Errors.UNSUPPORTED_VERSION.exception())
                    .completed();
        }
```

We can argue about making each branch of the if condition call a different method e.g. `onSupportedRequest(apiKey, header, request, context)` vs `onUnsupportedRequest(apiKey, header, request, context)` but I think its worth pulling the check into this method.